module/os/linux/zfs/zpl_xattr.c

```diff
@@ -706,19 +704,33 @@ static int
 __zpl_xattr_user_get(struct inode *ip, const char *name,
     void *value, size_t size)
 {
-	char *xattr_name;
+	boolean_t compat = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_COMPAT);
+	boolean_t fallback = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_FALLBACK);
 	int error;
 	/* xattr_resolve_name will do this for us if this is defined */
 #ifndef HAVE_XATTR_HANDLER_NAME
 	if (strcmp(name, "") == 0)
 		return (-EINVAL);
 #endif
+	if (ZFS_XA_NS_PREFIX_FORBIDDEN(name))
+		return (-EINVAL);
 	if (!(ITOZSB(ip)->z_flags & ZSB_XATTR))
 		return (-EOPNOTSUPP);
 
-	xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
-	error = zpl_xattr_get(ip, xattr_name, value, size);
-	kmem_strfree(xattr_name);
+	/*
+	 * Try to look up the name without the namespace prefix first for
+	 * compatibility with xattrs from other platforms.  If that fails,
+	 * try again with the namespace prefix if fallback is allowed.
+	 */
```

Comment:
It seems to me what we really want is a deterministic way to know if an xattr was written before or after the `xattr_compat` feature was enabled.  If it was written before then we need to invoke the fallback logic since it may or may not be prefixed.  However, if it was written after there's no ambiguity and thus no need for any fallback logic.

One way to do this would be to automatically upgrade all of xattrs attached to a znode when it's first accessed if 1) the `xattr_compat` feature is enabled and 2) and xattr upgraded bit isn't yet set.  There are available bits in the SA_ZPL_FLAGS system attribute we could reserve for this purpose.  All newly written xattrs could set this bit by default.   This would have the effect of gradually migrating all xattrs to the compatible format and the fallback logic would only be invoked once.  One nice this about this is then we shouldn't need a new fallback property.

> Allow choosing between cross-platform comparability and legacy Linux compatibility with a per-dataset property. This facilitates replication between hosts with different compatibility needs.

I believe what you're getting at here is send/recv compatibility since there currently is no one way to store a user xattr for all platforms.  This would allow a send stream to be generated which would be compatible with the xattr format used in older version of ZFS on Linux.

Reply:
A user xattr written before the xattr_compat feature is enabled *might* need the fallback, iff it was written on Linux. If it was written on any other platform, the fallback is not needed. We don't know in ZFS which platform things were written on a priori, so the xattr_fallback property allows for the fallback to be explicitly opted out of for performance when you know the xattrs written before xattr_compat did not come from Linux.

This does seem potentially useful still. One of my open questions is how to automatically skip the fallback logic in cases where we can assert it is safe to do so. We also have to know the xattr_compat property was never set to xattr_compat=linux, but if there are flags we can leverage that might do the trick.

Send/recv compatibility is the purpose of the xattr_compat property, yes.

Reply:
> One of my open questions is how to automatically skip the fallback logic in cases where we can assert it is safe to do so.

My feeling is it would be a nice simplification if we could also remove the need for the `xattr_compat` property. Then we could always write the data in the xattr_compat on-disk format and never need the fallback logic for upgraded znodes.  You'd just always pay the upgrade cost on first access after enabling the feature and then never again (except for read-only filesystems and snapshots).  We'd just need to find somewhere to store the bit indicating the znode's xattrs were upgraded. The SA_ZPL_FLAGS system attribute is one possible easy option.  We could probably also devine it based on birth times in the block pointers storing the xattrs (sa and directory), but I suspect we'd need to violate some layering to get that information so it may not be a great option.  If we thought about it some more I'm sure we could come up with other alternatives. 

The only catch I think is send/recv.  It might be possible to add a new `zfs send` flag which could be used to rewrite the xattr names in the send stream when its generated.  Or alternately, and perhaps much better, we could always generate the xattr_compat format and the `zstream` command could be taught to translate existing streams in to either format.  This way you could still generate a stream normally and then after the fact tailor it for target systems running older software.

Reply:
I'm not entirely on board with the idea of rewriting xattrs on znode access. That seems like it will potentially create unexpected load spikes that have no obvious explanation. We may potentially stat a very large number of files at once (such as for listdir), and that will create a lot of xattr rewrites. With xattr=sa there is still anIt's only once per file, but there's also the issue of read-only datasets such as snapshots. Samba exposes previous versions of files saved in snapshots using the .zfs dir. The snapshot data can't be rewritten, so we'll need all the same mechanisms for ensuring access to xattrs in those datasets anyway.

The properties allow users of both platforms to control the performance/compatibility tradeoffs they are willing to make. Without them, we leave cases where we cannot make safe assumptions about compatibility and must sacrifice performance. These cases are the very cross-platform migration situations we're trying to facilitate. Given an existing pool to be migrated, all xattrs were written before the feature was enabled, so we can't make any assumption in ZFS about the way the xattr names are stored. Upgrading the pool before the migration doesn't change the data already on disk.

* FreeBSD -> Linux
    The user knows this pool came from FreeBSD. We cannot determine this in ZFS, but the user can tell us by setting xattr_compat=all. The user also knows they have never written xattrs in this pool from Linux, so they don't need a fallback mechanism. We cannot determine this in ZFS, but the user can tell us by setting xattr_fallback=off, allowing us to avoid a performance regression on Linux.

* Linux -> FreeBSD
    The user knows this pool came from Linux. ZFS can't know this except by the xattr_compat property. The user also knows they have never written xattrs in this pool from another platform, so they are all in xattr_compat=linux format. ZFS can't know this except by xattr_fallback=off, allowing us to avoid a performance regression on FreeBSD.

The properties also allow us to avoid performance regression on Linux when upgrading existing pools that are only used on Linux. Setting xattr_compat=linux and xattr_fallback=off means the xattr format will remain unchanged and we will not need to pay the cost for compat fallbacks.

The above situations also apply when receiving datasets from another pool rather than importing a whole pool from another platform. The properties allow for separate datasets to have different formats, so for example one pool can receive streams from a mix of different systems for backup purposes, without requiring the user to remember to perform some extra translation step on their streams.

There are surely ways we could automatically ignore xattr_fallback=on and skip fallback when it can be guaranteed safe. I don't think we can avoid introducing both properties entirely without penalizing users switching between platforms. We have already experienced xattr operations being a bottleneck for Samba, which uses primarily xattrs to implement the MS-FSA layer. Making sure we have the options to avoid additional overhead in these operations is important.